app/src/main/kotlin/cz/covid19cz/erouska/crypto/keystore.kt

```diff
@@ -0,0 +1,93 @@
+package cz.covid19cz.erouska.crypto
+
+import android.content.Context
+import android.security.KeyPairGeneratorSpec
+import cz.covid19cz.erouska.db.SharedPrefsRepository
+import java.math.BigInteger
+import java.nio.charset.Charset
+import java.security.*
+import java.util.*
+import javax.crypto.Cipher
+import javax.security.auth.x500.X500Principal
+
+private const val ENCRYPTION_ALGORITHM = "RSA/ECB/PKCS1Padding"
+
+// Must be <= 256 bytes
+private const val DATABASE_KEY_SIZE = 32
+
+const val DATABASE_KEY_ALIAS = "db-key"
+
+fun getDatabaseEncryptionKey(prefsRepository: SharedPrefsRepository, context: Context): ByteArray? {
+    val key = prefsRepository.getDbEncryptedKey()
+    val secretKey = getAsymmetricStoreKey(context, DATABASE_KEY_ALIAS)
+
+    if (key == null) {
+        val randomBytes = generateRandomBytes()
+        val encryptedKey = encryptInStore(randomBytes, secretKey.secretKey) ?: randomBytes
+        prefsRepository.setDbEncryptedKey(encodeBytes(encryptedKey))
+        return randomBytes
+    }
+    return decryptInStore(decodeBytes(key), secretKey.secretKey)
+}
+
+private fun encodeBytes(bytes: ByteArray): String = String(bytes, Charset.forName("UTF-8"))
+private fun decodeBytes(string: String): ByteArray = string.toByteArray(Charset.forName("UTF-8"))
+
+/**
+ * Create or load an asymmetric secret key from the Android KeyStore.
+ */
+private fun getAsymmetricStoreKey(context: Context, alias: String): KeyStore.SecretKeyEntry {
+    val keyStore = KeyStore.getInstance("AndroidKeyStore").apply {
+        load(null)
```

Comment:
What exactly happens when keystore is not found?

Reply:
No idea, we can wrap it in try/catch and not encrypt if anything fails. But it should be on all devices from 4.3 up.

Reply:
I don't trust Android manufacturers :) We should gracefully fallback if anything fails.